pkg/uprobetracer/bpf/usdt_helper.bpf.c

```diff
@@ -0,0 +1,369 @@
+		u64 buf;                                             \
+		asm("%[res] = *(u64 *)(%[base] + %[offset])"         \
+		    : [res] "=r"(buf)                                \
+		    : [base] "r"(ctx),                               \
+		      [offset] "i"(offsetof(struct pt_regs, field)), \
+		      "m"(*ctx));                                    \
+		buf;                                                 \
+	})
+
+static __always_inline __u64 usdt_get_register_val(const struct pt_regs *ctx,
+						   unsigned reg_id)
+{
+	switch (reg_id) {
+	case USDT_ARG_REG_ZERO:
+		return 0;
+	case USDT_ARG_REG_R15:
+		return GET_CTX_FIELD(r15);
+	case USDT_ARG_REG_R14:
+		return GET_CTX_FIELD(r14);
+	case USDT_ARG_REG_R13:
+		return GET_CTX_FIELD(r13);
+	case USDT_ARG_REG_R12:
+		return GET_CTX_FIELD(r12);
+	case USDT_ARG_REG_BP:
+		return GET_CTX_FIELD(bp);
+	case USDT_ARG_REG_BX:
+		return GET_CTX_FIELD(bx);
+	case USDT_ARG_REG_R11:
+		return GET_CTX_FIELD(r11);
+	case USDT_ARG_REG_R10:
+		return GET_CTX_FIELD(r10);
+	case USDT_ARG_REG_R9:
+		return GET_CTX_FIELD(r9);
+	case USDT_ARG_REG_R8:
+		return GET_CTX_FIELD(r8);
+	case USDT_ARG_REG_AX:
+		return GET_CTX_FIELD(ax);
+	case USDT_ARG_REG_CX:
+		return GET_CTX_FIELD(cx);
+	case USDT_ARG_REG_DX:
+		return GET_CTX_FIELD(dx);
+	case USDT_ARG_REG_SI:
+		return GET_CTX_FIELD(si);
+	case USDT_ARG_REG_DI:
+		return GET_CTX_FIELD(di);
+	case USDT_ARG_REG_ORIG_AX:
+		return GET_CTX_FIELD(orig_ax);
+	case USDT_ARG_REG_IP:
+		return GET_CTX_FIELD(ip);
+	case USDT_ARG_REG_CS:
+		return GET_CTX_FIELD(cs);
+	case USDT_ARG_REG_FLAGS:
+		return GET_CTX_FIELD(flags);
+	case USDT_ARG_REG_SP:
+		return GET_CTX_FIELD(sp);
+	case USDT_ARG_REG_SS:
+		return GET_CTX_FIELD(ss);
+	default:
+		return 0;
+	}
+}
+
+#else
+#error "USDT argument support is not supported in the current architecture."
+#endif
+
+/* Some more complex USDT parameters may require multiple registers */
+/* We use 64 bits for future scalability */
+struct __usdt_argument {
+	unsigned magic : 4;
+	enum USDT_ARG_TYPE type : 1;
+	enum USDT_ARG_LENGTH length : 3;
+	enum USDT_ARG_REG reg : 8;
+	int offset : 16;
+	unsigned _reserved : 32;
+};
+
+struct __usdt_arguments {
+	struct __usdt_argument arguments[USDT_ARG_MAX_COUNT];
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(key_size, sizeof(u32));
+	__uint(value_size, sizeof(u64));
+	__uint(max_entries, 1);
+} __usdt_args_buffer SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(key_size, sizeof(u64));
+	__uint(value_size, sizeof(struct __usdt_arguments));
+	__uint(max_entries, 1024);
+} __usdt_args_info SEC(".maps");
+
+/* Returns true if the argument value is stored in the __usdt_args_buffer map successfully, otherwise returns false */
+/* The __usdt_args_buffer map will be reused, so users need to store the value before next call */
+SEC("freplace/__usdt_get_argument")
+bool __usdt_get_argument(struct pt_regs *ctx, u64 arg_idx)
```

Comment:
This function is used to get the argument. Because ebpf extension can only return one scalar, and we at least need to return a pair<bool success, long value>, the return value have to be stored in an ebpf map.
Here the return value is the "bool success".